src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+    }
+}
+
+mod internal {
+
+    use super::*;
+    use bytes::Bytes;
+    use std::io::{self, ErrorKind};
+    use std::path::{Path, PathBuf};
+
+    use futures::{try_ready, Async, AsyncSink, Future, Poll, Sink, StartSend};
+    use tokio::codec::{BytesCodec, FramedWrite};
+    use tokio::fs::file::{File, OpenFuture};
+    use tokio::fs::OpenOptions;
+
+    use tracing::field;
+
+    pub struct FileSink {
+        pub path: PathBuf,
+        state: FileSinkState,
+    }
+
+    enum FileSinkState {
+        Disconnected,
+        OpeningFile(OpenFuture<PathBuf>),
+        FileProvided(FramedWrite<File, BytesCodec>),
+    }
+
+    impl FileSinkState {
+        fn init(path: PathBuf) -> Self {
+            debug!(message = "opening", file = ?path);
+            let mut options = OpenOptions::new();
+            options.create(true).append(true);
+
+            FileSinkState::OpeningFile(options.open(path))
+        }
+    }
+
+    impl FileSink {
+        pub fn new(path: PathBuf) -> Self {
+            Self {
+                path: path.clone(),
+                state: FileSinkState::init(path),
+            }
+        }
+
+        pub fn new_with_encoding(path: &Path, encoding: Option<Encoding>) -> EmbeddedFileSink {
+            let sink = FileSink::new(path.to_path_buf())
+                .sink_map_err(|err| error!("Terminating the sink due to error: {}", err))
+                .with(move |event| Self::encode_event(event, &encoding));
+
+            Box::new(sink)
+        }
+
+        pub fn poll_file(&mut self) -> Poll<&mut FramedWrite<File, BytesCodec>, io::Error> {
+            loop {
+                match self.state {
+                    FileSinkState::Disconnected => return Err(disconnected()),
+
+                    FileSinkState::FileProvided(ref mut sink) => return Ok(Async::Ready(sink)),
+
+                    FileSinkState::OpeningFile(ref mut open_future) => match open_future.poll() {
+                        Ok(Async::Ready(file)) => {
+                            debug!(message = "provided", file = ?file);
+                            self.state = FileSinkState::FileProvided(FramedWrite::new(
+                                file,
+                                BytesCodec::new(),
+                            ));
+                        }
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Err(err) => {
+                            self.state = FileSinkState::Disconnected;
+                            return Err(err);
+                        }
+                    },
+                }
+            }
+        }
+
+        fn encode_event(event: Event, encoding: &Option<Encoding>) -> Result<Bytes, ()> {
+            let log = event.into_log();
+
+            let result = match (encoding, log.is_structured()) {
+                (&Some(Encoding::Json), _) | (_, true) => serde_json::to_vec(&log.all_fields())
+                    .map_err(|e| panic!("Error encoding: {}", e)),
```

Comment:
If somehow this event fails to encode this will actually crash the whole sink. Here is an example of how we handle it in the kinesis sink https://github.com/timberio/vector/blob/master/src/sinks/aws_kinesis_streams.rs#L196

Reply:
Actually, both examples panic :) But I see new cool method `unflatten`, which should be useful here also.